File: tests/blind-diff/detector.ts
// Independent (blind) TypeScript detector for admission_result.v0
// Built from SPEC.md §12–§16 + schemas + vector .input ONLY.
// Run: npx tsx detector.ts   (or: bun detector.ts)
//
// Modeling note: the vectors pre-abstract raw artifacts into semantic
// `evidence` flags (complete_comparable_interval, witness_live, observation
// positions, deadlines, authority class, explicit `violations` map, etc.).
// A fuller detector would recompute SHA-256 roots and verify Ed25519
// signatures; here those outcomes are supplied via evidence, so we drive the
// admission evaluation from the evidence exactly as the vector encodes it and
// treat artifacts as well-formed unless a violation is asserted. // SPEC §16 (vectors encode outcomes)

import * as fs from "fs";
import * as path from "path";

// ---- Reason-code ordering metadata (SPEC §13.4 check-ordinal + intra-check order) ----
// One reason code belongs to exactly one check ordinal (SPEC §13.5).
const REASON_ORDER: Record<string, { check: number; intra: number }> = {};
const CHECK_CODES: Record<number, string[]> = {
  1: ["profile_or_artifact_malformed"],
  2: ["issuance_intent_malformed"],
  3: [
    "witness_receipt_root_malformed",
    "witness_receipt_root_mismatch",
    "unsupported_witness_receipt_signature_profile",
    "malformed_witness_receipt_signature",
    "invalid_witness_receipt_signature",
    "accepted_record_ref_mismatch",
  ],
  4: ["missing_coverage_binding"],
  5: ["coverage_source_authority_not_allowed"],
  6: ["duplicate_coverage_binding"],
  7: ["independent_completeness_unproven"],
  8: ["issuance_stream_fork", "sequence_gap", "predecessor_ref_mismatch"],
  9: ["issuance_position_inversion"],
  10: ["accepted_record_not_in_witness_log"],
  11: [
    "witness_checkpoint_root_malformed",
    "witness_checkpoint_root_mismatch",
    "unsupported_witness_checkpoint_signature_profile",
    "malformed_witness_checkpoint_signature",
    "invalid_witness_checkpoint_signature",
    "witness_checkpoint_chain_discontinuous",
    "witness_append_only_consistency_unproven",
  ],
  12: ["witness_stalled"],
  13: ["witness_equivocation"],
  14: ["terminal_to_intent_mismatch"],
  15: ["late_skip", "resolution_overdue", "late_resolution"],
  16: ["publication_overdue", "late_publication"],
  17: ["as_of_scope_malformed", "as_of_scope_mismatch"],
};
for (const [c, codes] of Object.entries(CHECK_CODES)) {
  codes.forEach((code, i) => (REASON_ORDER[code] = { check: Number(c), intra: i }));
}

// SPEC §13.6.D — malformed-state finding set
const MALFORMED_SET = new Set([
  "profile_or_artifact_malformed",
  "issuance_intent_malformed",
  "witness_receipt_root_malformed",
  "malformed_witness_receipt_signature",
  "witness_checkpoint_root_malformed",
  "malformed_witness_checkpoint_signature",
  "as_of_scope_malformed",
]);

type AdmissionResult = {
  schema: string;
  coverage_profile_ref: string;
  accepted_record_schema: string;
  accepted_record_ref: string;
  witness_receipt_ref: string;
  as_of: { checkpoint_ref: string; log_size: number };
  predicate_evaluation_state: string;
  predicate_verdict: string | null;
  admission_evaluation_state: string;
  admission_verdict: string | null;
  resolution_progress: string;
  resolution_timing: string;
  publication_progress: string;
  publication_timing: string;
  findings: string[];
  primary_reason_code: string | null;
  unverifiable_checks: number[];
};

function detect(input: any): AdmissionResult {
  const A = input.artifacts;
  const ev = input.evidence;
  const evp = input.evaluation_point;
  const asOfPos: number = evp.position; // as_of position on the basis
  const complete: boolean = ev.complete_comparable_interval === true;

  const findings: string[] = [];
  const unverifiable = new Set<number>();

  const res = ev.resolution || {};
  const pub = ev.publication || {};
  const obs = ev.external_observation || {};
  const terminalPresent = res.terminal_present === true;

  // ---------- Predicate (SPEC §12.2, §14) ----------
  // Predicate is the substantive gate; independent of admission. Terminal present
  // (a computed result exists) => predicate evaluated/valid; pending => not_evaluated.
  // JUDGMENT CALL: no evidence field carries the predicate result; inferred from
  // terminal presence per matrix A/B/C (not_evaluated) vs D–K (evaluated/valid).
  let predicate_evaluation_state = "not_evaluated";
  let predicate_verdict: string | null = null;
  if (terminalPresent) {
    predicate_evaluation_state = "evaluated";
    predicate_verdict = "valid";
  }

  // ---------- Resolution progress / timing (SPEC §5, §11, §14) ----------
  let resolution_progress: string;
  let resolution_timing: string;
  if (terminalPresent) {
    resolution_progress = "resolved"; // SPEC §11
    if (res.terminal_position <= res.deadline_position) {
      resolution_timing = "on_time"; // SPEC §11 event<=deadline
    } else {
      resolution_timing = "late"; // SPEC §11 terminal after deadline => late + late_resolution
      addFinding(findings, "late_resolution");
      // SPEC §11: "A later terminal MUST NOT erase the earlier overdue interval."
      // Co-occurrence: when the overdue interval was previously proven, both findings
      // co-occur (resolution_overdue precedes late_resolution within check 15). // SPEC §13.4 check15 order
      if (res.overdue_interval_previously_proven === true) {
        addFinding(findings, "resolution_overdue");
      }
    }
  } else {
    resolution_progress = "pending"; // SPEC §11
    if (res.deadline_position == null || !complete) {
      // Cannot prove deadline state -> not_started + check 15 unverifiable. // SPEC §14.C, §13.1
      resolution_timing = "not_started";
      unverifiable.add(15);
    } else if (asOfPos <= res.deadline_position) {
      resolution_timing = "not_started"; // SPEC §5 not_started before deadline
    } else {
      resolution_timing = "overdue"; // SPEC §11 no terminal, as_of>deadline, complete interval
      addFinding(findings, "resolution_overdue");
    }
  }

  // ---------- Witness liveness (check 12, SPEC §10) ----------
  if (ev.witness_live === false) {
    if (complete) addFinding(findings, "witness_stalled"); // SPEC §10, §14.B
    else unverifiable.add(12); // SPEC §10 endpoint unavailability => unverifiable, not proven
  }

  // ---------- Publication progress / timing (SPEC §11, §14) ----------
  let publication_progress: string;
  let publication_timing: string;
  if (!terminalPresent) {
    // Publication window starts at witness acceptance of the terminal; none yet. // SPEC §11
    publication_progress = "not_started";
    publication_timing = "not_started";
  } else if (obs.observed === true) {
    publication_progress = "externally_observed"; // SPEC §9, §11
    if (obs.position <= pub.deadline_position) {
      publication_timing = "on_time"; // SPEC §11
    } else {
      publication_timing = "late"; // SPEC §11 later observation after deadline => late_publication
      addFinding(findings, "late_publication");
    }
  } else {
    const pubOverdue = complete && asOfPos > pub.deadline_position;
    if (pubOverdue) {
      publication_timing = "overdue"; // SPEC §11 publication_overdue
      addFinding(findings, "publication_overdue");
      // Progress reflects actual publication state at the overdue point. // SPEC §11 G1/G2
      publication_progress = pub.checkpoint_published === true ? "published" : "accepted_by_witness";
    } else {
      publication_timing = "not_started";
      // JUDGMENT CALL: no explicit "accepted_by_witness" evidence flag. Matrix E
      // (terminal resolved, not published, deadline not passed) => not_started, so
      // absent checkpoint publication we report not_started unless overdue forces
      // the accepted_by_witness intermediate.
      publication_progress = pub.checkpoint_published === true ? "published" : "not_started";
      if (!complete) unverifiable.add(12); // SPEC §14.F external observation unverifiable
    }
  }

  // ---------- Authority / independent completeness (check 7, SPEC §4, §13) ----------
  const auth = ev.authority || {};
  if (auth.independent_completeness === "proven_unavailable") {
    addFinding(findings, "independent_completeness_unproven"); // SPEC §13 check7, §14.K1
  } else if (auth.independent_completeness === "insufficient_evidence") {
    unverifiable.add(7); // SPEC §14.K2, §13.1
  }

  // ---------- Equivocation (check 13, SPEC §J) ----------
  if (Array.isArray(ev.incompatible_checkpoint_pair)) {
    addFinding(findings, "witness_equivocation"); // SPEC §14.J check13
  }

  // ---------- Explicit violations injection (co-occurrence findings-ordering vector) ----------
  // JUDGMENT CALL: the co-multiple-violations vector encodes proven findings via a
  // `violations` map rather than raw crypto/timing artifacts; treat each true key as
  // a proven finding, and align the corresponding timing literal for consistency.
  if (ev.violations && typeof ev.violations === "object") {
    for (const [code, on] of Object.entries(ev.violations)) {
      if (on === true && REASON_ORDER[code]) addFinding(findings, code);
    }
    if (ev.violations.late_resolution === true) {
      resolution_progress = "resolved";
      resolution_timing = "late";
    }
    if (ev.violations.late_publication === true) {
      publication_timing = "late";
    }
  }

  // ---------- Assemble findings ordering (SPEC §13.2) ----------
  findings.sort((a, b) => {
    const ma = REASON_ORDER[a];
    const mb = REASON_ORDER[b];
    if (ma.check !== mb.check) return ma.check - mb.check;
    return ma.intra - mb.intra;
  });

  const unverifiable_checks = Array.from(unverifiable).sort((a, b) => a - b); // SPEC §13.1

  // ---------- Admission evaluation state / verdict (SPEC §13.6) ----------
  let admission_evaluation_state: string;
  let admission_verdict: string | null;
  if (findings.length > 0) {
    const isMalformed = findings.some((f) => MALFORMED_SET.has(f));
    if (isMalformed) {
      admission_evaluation_state = "malformed"; // SPEC §13.6.D
      admission_verdict = null;
    } else {
      admission_evaluation_state = "evaluated"; // SPEC §13.6.B
      admission_verdict = "invalid";
    }
  } else if (unverifiable_checks.length > 0) {
    admission_evaluation_state = "unverifiable"; // SPEC §13.6.C
    admission_verdict = null;
  } else if (
    resolution_progress === "resolved" &&
    publication_progress === "externally_observed" &&
    resolution_timing === "on_time" &&
    publication_timing === "on_time"
  ) {
    admission_evaluation_state = "evaluated"; // SPEC §13.6.A (all checks passed) — matrix D/I
    admission_verdict = "valid";
  } else {
    admission_evaluation_state = "not_evaluated"; // SPEC §14.A pending, nothing proven/unverifiable
    admission_verdict = null;
  }

  const primary_reason_code = findings.length > 0 ? findings[0] : null; // SPEC §13.3

  // ---------- Identity refs (SPEC §12.1) ----------
  // JUDGMENT CALL: accepted record is the terminal (issuance_result_commitment.v0) when
  // resolved, else the issuance intent (issuance_record.v0). witness_receipt_ref is the
  // matching receipt's stored root.
  let accepted_record_schema: string;
  let accepted_record_ref: string;
  if (terminalPresent) {
    accepted_record_schema = "issuance_result_commitment.v0";
    accepted_record_ref = A.terminal_commitment.issuance_result_root;
  } else {
    accepted_record_schema = "issuance_record.v0";
    accepted_record_ref = A.issuance_intent.issuance_root;
  }
  const receipt = (A.witness_receipts || []).find(
    (r: any) => r.accepted_record_schema === accepted_record_schema
  );
  const witness_receipt_ref = receipt ? receipt.witness_receipt_root : "";

  return {
    schema: "admission_result.v0",
    coverage_profile_ref: A.coverage_profile.coverage_profile_root,
    accepted_record_schema,
    accepted_record_ref,
    witness_receipt_ref,
    as_of: { checkpoint_ref: evp.as_of.checkpoint_ref, log_size: evp.as_of.log_size },
    predicate_evaluation_state,
    predicate_verdict,
    admission_evaluation_state,
    admission_verdict,
    resolution_progress,
    resolution_timing,
    publication_progress,
    publication_timing,
    findings,
    primary_reason_code,
    unverifiable_checks,
  };
}

function addFinding(list: string[], code: string) {
  if (!list.includes(code)) list.push(code); // uniqueItems (SPEC §13.2)
}

// ---------- Runner ----------
const VECTOR_DIR = path.join(__dirname, "vectors");
const order = [
  "a", "b", "c", "d", "e", "f", "g1", "g2", "h", "i", "j", "k1", "k2",
  "co-equivocation-prefix", "co-late-terminal-after-overdue", "co-multiple-violations",
];

if (require.main === module) {
  const mode = process.argv[2] || "emit";
  const out: Record<string, AdmissionResult> = {};
  for (const name of order) {
    const vec = JSON.parse(fs.readFileSync(path.join(VECTOR_DIR, name + ".json"), "utf8"));
    out[vec.case_id] = detect(vec.input);
  }
  if (mode === "emit") {
    console.log(JSON.stringify(out, null, 2));
  }
}

export { detect };

File: tests/blind-diff/detector.py
#!/usr/bin/env python3
"""
Independent blind detector for UNANCHORED_ISSUANCE_WITNESS_V0.
Built from SPEC.md + schemas/* ONLY. No vector .expected was read while
authoring this file. Maps a vector .input -> admission_result.v0.

Each nontrivial rule carries a `# SPEC section` citation so the diff is auditable.

The vector `evidence` object is a pre-digested statement of proven facts
(complete_comparable_interval, witness_live, external_observation, resolution,
publication, authority, and optional incompatible_checkpoint_pair / violations).
The detector consumes that digest; it does not re-run SHA-256/Ed25519, which the
spec leaves to the crypto profile doc. Structural crypto findings (check 3 / 11)
are only produced when a vector injects them via evidence.violations.
"""

import json
import sys

# ---------------------------------------------------------------------------
# SPEC 13.4 : closed reason-code enum, grouped by admission-check ordinal and
# with the normative within-check order. (check_ordinal, within_index)
# ---------------------------------------------------------------------------
CODE_TABLE = {
    # Check 1
    "profile_or_artifact_malformed": (1, 0),
    # Check 2
    "issuance_intent_malformed": (2, 0),
    # Check 3 (order pinned in 13.4 / 13.5)
    "witness_receipt_root_malformed": (3, 0),
    "witness_receipt_root_mismatch": (3, 1),
    "unsupported_witness_receipt_signature_profile": (3, 2),
    "malformed_witness_receipt_signature": (3, 3),
    "invalid_witness_receipt_signature": (3, 4),
    "accepted_record_ref_mismatch": (3, 5),
    # Check 4
    "missing_coverage_binding": (4, 0),
    # Check 5
    "coverage_source_authority_not_allowed": (5, 0),
    # Check 6
    "duplicate_coverage_binding": (6, 0),
    # Check 7
    "independent_completeness_unproven": (7, 0),
    # Check 8
    "issuance_stream_fork": (8, 0),
    "sequence_gap": (8, 1),
    "predecessor_ref_mismatch": (8, 2),
    # Check 9
    "issuance_position_inversion": (9, 0),
    # Check 10
    "accepted_record_not_in_witness_log": (10, 0),
    # Check 11
    "witness_checkpoint_root_malformed": (11, 0),
    "witness_checkpoint_root_mismatch": (11, 1),
    "unsupported_witness_checkpoint_signature_profile": (11, 2),
    "malformed_witness_checkpoint_signature": (11, 3),
    "invalid_witness_checkpoint_signature": (11, 4),
    "witness_checkpoint_chain_discontinuous": (11, 5),
    "witness_append_only_consistency_unproven": (11, 6),
    # Check 12
    "witness_stalled": (12, 0),
    # Check 13
    "witness_equivocation": (13, 0),
    # Check 14
    "terminal_to_intent_mismatch": (14, 0),
    # Check 15
    "late_skip": (15, 0),
    "resolution_overdue": (15, 1),
    "late_resolution": (15, 2),
    # Check 16
    "publication_overdue": (16, 0),
    "late_publication": (16, 1),
    # Check 17
    "as_of_scope_malformed": (17, 0),
    "as_of_scope_mismatch": (17, 1),
}

# SPEC 13.6 D : findings that force admission_evaluation_state = malformed
MALFORMED_CODES = {
    "profile_or_artifact_malformed",
    "issuance_intent_malformed",
    "witness_receipt_root_malformed",
    "malformed_witness_receipt_signature",
    "witness_checkpoint_root_malformed",
    "malformed_witness_checkpoint_signature",
    "as_of_scope_malformed",
}


def order_findings(codes):
    """SPEC 13.2 : sort by check ordinal, then within-check normative order."""
    uniq = list(dict.fromkeys(codes))  # dedupe, uniqueItems (schema)
    return sorted(uniq, key=lambda c: CODE_TABLE[c])


def evaluate(vin):
    art = vin["artifacts"]
    ev = vin["evidence"]
    ep = vin["evaluation_point"]

    P = ep["position"]                      # as_of position on the declared basis
    res = ev["resolution"]
    pub = ev["publication"]
    obs = ev["external_observation"]
    auth = ev["authority"]
    complete = ev["complete_comparable_interval"]   # SPEC 5/11: overdue/observation-absence
                                                    # require a proven complete interval
    terminal_present = res["terminal_present"]

    findings = []          # collected reason-code strings (proven violations)
    unverifiable = set()   # SPEC 13.1 : integers 1..17

    # -----------------------------------------------------------------
    # Identity refs (SPEC 12.1). The "accepted record under evaluation" is the
    # computation *result* (terminal commitment) once one exists, else the
    # issuance intent. [JUDGMENT CALL — spec does not name which record the
    # admission view binds to when both exist; see report.]
    # -----------------------------------------------------------------
    coverage_profile_ref = art["coverage_profile"]["coverage_profile_root"]
    intent_root = art["issuance_intent"]["issuance_root"]
    result_root = art["terminal_commitment"]["issuance_result_root"]

    def receipt_root_for(schema):
        for r in art["witness_receipts"]:
            if r["accepted_record_schema"] == schema:
                return r["witness_receipt_root"]
        return None

    if terminal_present:
        accepted_record_schema = "issuance_result_commitment.v0"
        accepted_record_ref = result_root
        witness_receipt_ref = receipt_root_for("issuance_result_commitment.v0")
    else:
        accepted_record_schema = "issuance_record.v0"
        accepted_record_ref = intent_root
        witness_receipt_ref = receipt_root_for("issuance_record.v0")

    as_of = {
        "checkpoint_ref": ep["as_of"]["checkpoint_ref"],
        "log_size": ep["as_of"]["log_size"],
    }

    # -----------------------------------------------------------------
    # Check 7 — independent completeness (SPEC 13, 13.4 check 7; matrix K1/K2)
    # proven_unavailable -> proven finding; insufficient_evidence -> unverifiable.
    # -----------------------------------------------------------------
    ic = auth.get("independent_completeness")
    if ic == "proven_unavailable":
        findings.append("independent_completeness_unproven")
    elif ic == "insufficient_evidence":
        unverifiable.add(7)
    # "proven" -> check 7 passes

    # -----------------------------------------------------------------
    # Check 12 — external observation / witness liveness (SPEC 10, 12; matrix B)
    # witness_stalled only with a proven complete comparable interval; endpoint
    # unavailability alone (incomplete interval) is unverifiable, not a finding.
    # -----------------------------------------------------------------
    if not ev["witness_live"]:
        if complete:
            findings.append("witness_stalled")   # SPEC 10: missed cadence proven
        else:
            unverifiable.add(12)                 # SPEC 10/13.1: unavailability != proven

    # -----------------------------------------------------------------
    # Check 13 — equivocation status (SPEC 14.J, 13.4 check 13)
    # An incompatible checkpoint pair is a proven check-13 violation. This is
    # distinct from append-only consistency merely being unproven (SPEC 13.5).
    # -----------------------------------------------------------------
    if "incompatible_checkpoint_pair" in ev:
        findings.append("witness_equivocation")

    # -----------------------------------------------------------------
    # Check 15 — resolution timing / progress (SPEC 5, 11)
    # -----------------------------------------------------------------
    res_deadline = res.get("deadline_position")
    if terminal_present:
        resolution_progress = "resolved"          # SPEC 11 closed domain
        tpos = res["terminal_position"]
        if tpos <= res_deadline:                  # SPEC 11: event<=deadline => on_time
            resolution_timing = "on_time"
        else:
            resolution_timing = "late"            # SPEC 11: arrives after deadline
            findings.append("late_resolution")    # finding late_resolution (check 15)
    else:
        resolution_progress = "pending"
        if complete and res_deadline is not None and P > res_deadline:
            resolution_timing = "overdue"         # SPEC 11: no terminal, as_of>deadline,
            findings.append("resolution_overdue") # complete interval proven
        elif complete and res_deadline is not None:
            resolution_timing = "not_started"     # SPEC 5/11: deadline not yet passed
        else:
            # SPEC 11/matrix C: incomplete/incomparable interval -> cannot classify;
            # represented by evaluation state + unverifiable_checks, not a timing literal.
            resolution_timing = "not_started"
            unverifiable.add(15)

    # -----------------------------------------------------------------
    # Check 16 — publication timing + publication progress (SPEC 11)
    # publication_progress "not_started" means the terminal has not entered the
    # witness path (SPEC 11). No terminal => publication axis is not_started.
    # -----------------------------------------------------------------
    pub_deadline = pub.get("deadline_position")
    observed = obs["observed"]
    obs_pos = obs.get("position")
    if not terminal_present:
        publication_progress = "not_started"
        publication_timing = "not_started"
    else:
        if observed:
            publication_progress = "externally_observed"   # SPEC 9/11
            if obs_pos <= pub_deadline:
                publication_timing = "on_time"
            else:
                publication_timing = "late"                # SPEC 11: late external obs
                findings.append("late_publication")
        else:
            published = pub.get("checkpoint_published", False)
            deadline_passed = pub_deadline is not None and P > pub_deadline
            if deadline_passed and complete:
                publication_timing = "overdue"             # SPEC 11: proven publication overdue
                findings.append("publication_overdue")
                # SPEC 11: progress axis independent of timing
                publication_progress = "published" if published else "accepted_by_witness"
            elif deadline_passed and not complete:
                # SPEC 11/matrix F: cannot prove observation absence -> external
                # observation (check 12) unverifiable; keep closed timing literal.
                publication_timing = "not_started"
                unverifiable.add(12)
                publication_progress = "published" if published else "not_started"
            else:
                # deadline not passed (SPEC 11 / matrix E): publication has not
                # started for this evaluation.
                publication_timing = "not_started"
                publication_progress = "published" if published else "not_started"

    # -----------------------------------------------------------------
    # Explicit co-occurrence injection (SPEC 13.7 co-occurrence vectors):
    # a vector may declare the exact set of proven violations to test the
    # deterministic total ordering of findings (13.2). Union with primitives.
    # -----------------------------------------------------------------
    if "violations" in ev:
        for code, on in ev["violations"].items():
            if on:
                findings.append(code)

    # -----------------------------------------------------------------
    # SPEC 13.1 : an unverifiable check MUST NOT emit a finding for that check.
    # Drop any check from unverifiable_checks that also produced a proven finding.
    # -----------------------------------------------------------------
    finding_checks = {CODE_TABLE[c][0] for c in findings}
    unverifiable = {n for n in unverifiable if n not in finding_checks}

    # -----------------------------------------------------------------
    # Assemble findings ordering + primary_reason_code (SPEC 13.2, 13.3)
    # -----------------------------------------------------------------
    ordered = order_findings(findings)
    primary_reason_code = ordered[0] if ordered else None

    # -----------------------------------------------------------------
    # Admission evaluation-state / verdict (SPEC 13.6)
    # -----------------------------------------------------------------
    if ordered:
        if any(c in MALFORMED_CODES for c in ordered):
            admission_evaluation_state = "malformed"   # 13.6 D
            admission_verdict = None
        else:
            admission_evaluation_state = "evaluated"   # 13.6 B
            admission_verdict = "invalid"
    elif unverifiable:
        admission_evaluation_state = "unverifiable"    # 13.6 C
        admission_verdict = None
    else:
        # 13.6 A vs E: a resolved/determinable case is evaluated-valid; a still
        # -pending case with nothing proven and nothing unverifiable is
        # not_evaluated (matrix A).
        if terminal_present:
            admission_evaluation_state = "evaluated"   # 13.6 A
            admission_verdict = "valid"
        else:
            admission_evaluation_state = "not_evaluated"  # matrix A / 13.6 E
            admission_verdict = None

    # -----------------------------------------------------------------
    # Predicate dimension (SPEC 12.2 — independent of admission).
    # Predicate is evaluated once a terminal (computation result) exists; while
    # pending it is not_evaluated/null (SPEC 11 resolution_overdue also pins
    # not_evaluated/null). No vector encodes an invalid/unverifiable/malformed
    # predicate, so a resolved terminal yields valid. [JUDGMENT CALL — see report.]
    # -----------------------------------------------------------------
    if terminal_present:
        predicate_evaluation_state = "evaluated"
        predicate_verdict = "valid"
    else:
        predicate_evaluation_state = "not_evaluated"
        predicate_verdict = None

    return {
        "schema": "admission_result.v0",
        "coverage_profile_ref": coverage_profile_ref,
        "accepted_record_schema": accepted_record_schema,
        "accepted_record_ref": accepted_record_ref,
        "witness_receipt_ref": witness_receipt_ref,
        "as_of": as_of,
        "predicate_evaluation_state": predicate_evaluation_state,
        "predicate_verdict": predicate_verdict,
        "admission_evaluation_state": admission_evaluation_state,
        "admission_verdict": admission_verdict,
        "resolution_progress": resolution_progress,
        "resolution_timing": resolution_timing,
        "publication_progress": publication_progress,
        "publication_timing": publication_timing,
        "findings": ordered,
        "primary_reason_code": primary_reason_code,
        "unverifiable_checks": sorted(unverifiable),
    }


if __name__ == "__main__":
    d = json.load(open(sys.argv[1]))
    print(json.dumps(evaluate(d["input"]), indent=2))
